mapie/multi_label_classification.py

```diff
@@ -0,0 +1,738 @@
+        Optimal threshold according to alpha.
+
+    References
+    ----------
+    [1] Lihua Lei Jitendra Malik Stephen Bates, Anastasios Angelopoulos
+    and Michael I. Jordan. Distribution-free, risk-controlling prediction
+    sets. CoRR, abs/2101.02703, 2021.
+    URL https://arxiv.org/abs/2101.02703.39
+
+    [2] Angelopoulos, Anastasios N., Stephen, Bates, Adam, Fisch, Lihua,
+    Lei, and Tal, Schuster. "Conformal Risk Control." (2022).
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.multioutput import MultiOutputClassifier
+    >>> from sklearn.linear_model import LogisticRegression
+    >>> from mapie.multi_label_classification import MapieMultiLabelClassifier
+    >>> X_toy = np.arange(4).reshape(-1, 1)
+    >>> y_toy = np.stack([[1, 0, 1], [1, 0, 0], [0, 1, 1], [0, 1, 0]])
+    >>> clf = MultiOutputClassifier(LogisticRegression()).fit(X_toy, y_toy)
+    >>> mapie = MapieMultiLabelClassifier(estimator=clf).fit(X_toy, y_toy)
+    >>> _, y_pi_mapie = mapie.predict(X_toy, alpha=0.3)
+    >>> print(y_pi_mapie[:, :, 0])
+    [[ True False  True]
+     [ True False  True]
+     [False  True False]
+     [False  True False]]
+    """
+    valid_methods_ = ["crc", "rcps"]
+    n_lambdas = 100
+    fit_attributes = [
+        "single_estimator_",
+        "risks"
+    ]
+    sigma_init = .25
+
+    def __init__(
+        self,
+        estimator: Optional[ClassifierMixin] = None,
+        n_jobs: Optional[int] = None,
+        random_state: Optional[Union[int, np.random.RandomState]] = None,
+        verbose: int = 0
+    ) -> None:
+        self.estimator = estimator
+        self.n_jobs = n_jobs
+        self.random_state = random_state
+        self.verbose = verbose
+
+    def _check_parameters(self) -> None:
+        """
+        Perform several checks on input parameters.
+
+        Raises
+        ------
+        ValueError
+            If parameters are not valid.
+        """
+        check_n_jobs(self.n_jobs)
+        check_verbose(self.verbose)
+        check_random_state(self.random_state)
+
+    def _check_method(self) -> None:
+        """Check that the specified method is valid
+
+        Raises
+        ------
+        ValueError
+            Raise error if the name of the method is not
+            in ["crc", "rcps"]
+        """
+        if self.method not in self.valid_methods_:
+            raise ValueError(
+                "Invalid method. "
+                "Allowed values are 'crc' or 'rcps"
+            )
+
+    def _check_all_labelled(self, y: NDArray) -> None:
+        """Check that all observations have at least
+        one label
+
+        Parameters
+        ----------
+        y : NDArray of shape (n_samples, n_labels)
+            Labels of the observations.
+
+        Raises
+        ------
+        ValueError
+            Raise error if at least one observation
+            has no label.
+        """
+        if 0 in np.unique(y.sum(axis=1)):
```

Comment:
Becareful: 0 could be a label and then y.sum(axis=1) for this line
use (~np.isnan(y)).sum(axis=1) ?

Reply:
replace by
`` not (y.sum(axis=1) > 0).all() ``

Reply:
y has the shape (n_samples, n_labels) so even if 0 is a label I would have the following behavior:

`` y = 0 --> y = [1, 0, 0, 0]`` if there are 4 possible labels